### PR TITLE
Fix Token generator configuration on consumer level 

### PR DESCRIPTION
**Pull request by sameeragunarathne**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java [hunk 1]
```diff
@@ -25,6 +25,9 @@
   
   private List<String> grantTypes = new ArrayList<String>();
 
+
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java]: Remove extra new lines

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java [hunk 2]
```diff
@@ -61,7 +64,16 @@ public void setGrantTypes(List<String> grantTypes) {
...
 
-  
+
+  @ApiModelProperty(value = "")
+  @JsonProperty("token_type_extension")
+  public String getTokenType() {
+    return tokenType;
+  }
+
+  public void setTokenType(String tokenType) {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java]: Add a new line

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 3]
```diff
@@ -134,6 +134,10 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
                 String callbackUrl = validateAndSetCallbackURIs(updateRequest.getRedirectUris(), updateRequest.getGrantTypes());
                 appDTO.setCallbackUrl(callbackUrl);
             }
+
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: remove extra new line

File: components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java [hunk 4]
```diff
@@ -30,11 +30,13 @@
 import org.wso2.carbon.identity.oauth.stub.dto.OAuthRevocationResponseDTO;
 
 import java.rmi.RemoteException;
+import java.util.List;
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java]: Seemslike redundant import

File: components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java [hunk 5]
```diff
@@ -179,6 +181,22 @@ public OAuthTokenExpiryTimeDTO getOAuthTokenExpiryTimeDTO() throws RemoteExcepti
         return scopeValidators;
     }
 
+    /**
+     * Get the registered oauth token types from OAuth server configuration file.
+     *
+     * @return List of supported oauth token types
+     * @throws RemoteException exception occured during remote call
+     */
+    public String[] getSupportedTokenTypes() throws RemoteException {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java]: Add new line after the method declaration

File: components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java [hunk 6]
```diff
@@ -179,6 +181,22 @@ public OAuthTokenExpiryTimeDTO getOAuthTokenExpiryTimeDTO() throws RemoteExcepti
...
+        if (supportedTokenTypes == null) {
+            supportedTokenTypes = stub.getSupportedTokenTypes();
+            if (supportedTokenTypes == null) {
+                supportedTokenTypes = new String[0];
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java]: Shall we use ArrayUtils.EMPTY_STRING_ARRAY?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 7]
```diff
@@ -296,6 +308,7 @@ private void buildOAuthServerConfiguration() {
         // read supported grant types
         parseSupportedGrantTypesConfig(oauthElem);
 
+
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: remove unnecessary new line addition

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 8]
```diff
@@ -1798,6 +1823,121 @@ private void parseSupportedGrantTypesConfig(OMElement oauthConfigElem) {
...
+                    oauthTokenIssuer = (OauthTokenIssuer) clazz.newInstance();
+                    oauthTokenIssuer.setPersistAccessTokenAlias(
+                            supportedTokenIssuers.get(tokenType).isPersistAccessTokenAlias());
+                    oauthTokenIssuerMap.put(tokenType, oauthTokenIssuer);
+                    log.info("An instance of " + tokenIssuerDO.getTokenImplClass()
+                            + " is created for Identity OAuth token generation.");
+                } else {
+                    oauthTokenIssuer = oauthTokenIssuerMap.get(tokenType);
+                }
+            } catch (Exception e) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Is there a reason to catch generic Exception here?

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java (hunk 1 above)
Comment (sameeragunarathne) [reply components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/UpdateRequestDTO.java]: fixed with 0c8dfb39735118757027bcb9748516cebd8dd8c1

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 3 above)
Comment (sameeragunarathne) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: fixed with 0c8dfb39735118757027bcb9748516cebd8dd8c1

File: components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java (hunk 5 above)
Comment (sameeragunarathne) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java]: fixed with 0c8dfb39735118757027bcb9748516cebd8dd8c1

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 7 above)
Comment (sameeragunarathne) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: fixed with 0c8dfb39735118757027bcb9748516cebd8dd8c1

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 8 above)
Comment (sameeragunarathne) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: These are used to catch for class loading and instantiating issues. Similar implementations can be found in the same class such as https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java#L958

**Comment by denuwanthi** [comment]
It's better to add unit tests for the Token Generation configuration at service provider configuration level feature.

**Pull request merged**